### PR TITLE
Extend specs for IO#pread

### PR DESCRIPTION
**Pull request by herwinw**

### CODE REVIEW

File: core/io/pread_spec.rb [hunk 1]
```diff
@@ -26,11 +26,73 @@
...
+      @file.pread(maxlen, offset).should == "1234"
+    end
+
+    it "converts a buffer to String using to_str (but you cannot use this value)" do
+      buffer = mock('buffer')
+      buffer.should_receive(:to_str).and_return("foo")
+      @file.pread(0, 4, buffer).should == "foo"
+      buffer.should_not.is_a?(String)
+      buffer.to_str.should == "foo"
+    end
```
Comment (herwinw) [inline core/io/pread_spec.rb:79]: I'm not really sure about this one: this is the current behaviour of MRI, but this feels like an implementation detail/bug, since the value cannot be used.

Comment (herwinw) [reply core/io/pread_spec.rb]: The behaviour appears to be different between Linux and Windows as well.

File: core/io/pread_spec.rb [hunk 2]
```diff
@@ -26,11 +26,73 @@
...
+      @file.pread(maxlen, offset).should == "1234"
+    end
+
+    it "converts a buffer to String using to_str (but you cannot use this value)" do
+      buffer = mock('buffer')
+      buffer.should_receive(:to_str).at_least(1).and_return("foo")
+      @file.pread(0, 4, buffer).should == "foo"
+      buffer.should_not.is_a?(String)
+      buffer.to_str.should == "foo"
+    end
```
Comment (andrykonchin) [inline core/io/pread_spec.rb:79]: Both file content and a `#to_str` returned value are equal that seems confusing.

File: core/io/pread_spec.rb [hunk 3]
```diff
@@ -26,11 +26,73 @@
...
+      @file.pread(4, 0).should == "1234"
+    end
+
+    it "converts maxlen and offset to Integers using #to_int" do
+      maxlen = mock('maxlen')
+      maxlen.should_receive(:to_int).and_return(4)
+      offset = mock('offset')
+      offset.should_receive(:to_int).and_return(0)
+      @file.pread(maxlen, offset).should == "1234"
+    end
```
Comment (andrykonchin) [inline core/io/pread_spec.rb:71]: **suggestion:** It makes sense to have two separate test cases for each argument.

File: core/io/pread_spec.rb [hunk 4]
```diff
@@ -26,11 +26,73 @@
...
+      buffer.should == "foo"
+
+      @file.pread(0, 400, buffer)
+      buffer.should == "foo"
+    end
+
+    it "ignores the current offset" do
+      @file.pos = 3
+      @file.pread(4, 0).should == "1234"
+    end
```
Comment (andrykonchin) [inline core/io/pread_spec.rb:63]: **suggestion:** I would put this case a bit closer to the beginning near a case `"does not advance the file pointer"` as far as it is not an edge case but core behaviour of the method.

File: core/io/pread_spec.rb [hunk 5]
```diff
@@ -26,11 +26,73 @@
...
+
     it "does not advance the file pointer" do
       @file.pread(4, 0).should == "1234"
       @file.read.should == "1234567890"
     end
 
+    it "returns an empty string for maxlen = 0, regardless of the offset" do
+      @file.pread(0, 4).should == ""
+      @file.pread(0, 400).should == ""
+    end
```
Comment (andrykonchin) [inline core/io/pread_spec.rb:49]: **suggestion:** it's important to reflect in the test title that offset is not just ignored but even not validated and could be out of the file end boundary.

File: core/io/pread_spec.rb [hunk 6]
```diff
@@ -26,11 +26,73 @@
...
+      buffer.should_receive(:to_str).at_least(1).and_return("foo")
+      @file.pread(0, 4, buffer).should == "foo"
+      buffer.should_not.is_a?(String)
+      buffer.to_str.should == "foo"
+    end
+
+    it "expects Integer values for maxlen and offset" do
+      -> { @file.pread("maxlen", 0) }.should raise_error(TypeError)
+      -> { @file.pread(4, "offset") }.should raise_error(TypeError)
+    end
```
Comment (andrykonchin) [inline core/io/pread_spec.rb:84]: **suggestion:** it's an edge case when object doesn't respond to `#to_int` method, so I would reflect it in the test title (as well as in the test for non-String buffer).

File: core/io/pread_spec.rb [hunk 7]
```diff
@@ -26,11 +26,73 @@
...
+
+    it "does not support negative values for maxlen and offset" do
+      -> { @file.pread(-4, 0) }.should raise_error(ArgumentError)
+      -> { @file.pread(4, -1) }.should raise_error(Errno::EINVAL)
+    end
```
Comment (andrykonchin) [inline core/io/pread_spec.rb:89]: **suggestion:** it makes sense to have two separate test cases.

Also exception message are specified.

File: core/io/pread_spec.rb (hunk 2 above)
Comment (herwinw) [reply core/io/pread_spec.rb:94]: I think I fixed the issue here. My test was supposed to show that if you read into a buffer using `#to_str`, it totally ignored the buffer. Turns out I started this test by copy-pasting the wrong code, and instead of reading 4 bytes from offset 0, it read 0 bytes from offset 4. This behaviour is already covered in "does not reset the buffer when reading with maxlen = 0".

I've updated the test, it all makes sense now.

File: core/io/pread_spec.rb (hunk 6 above)
Comment (herwinw) [reply core/io/pread_spec.rb:98]: It turns out the error message in this case is different between linux and mac os (https://github.com/ruby/spec/actions/runs/6315519202/job/17148232880?pr=1079). What would be the best course of action here, matching `no implicit conversion` as the maximal shared string, or matching it with a Regex union of both messages?

And honestly, I think it's kind of weird that a non-int type for maxlen results in a different error message than offset, I doubt this is the intended behaviour.

Comment (andrykonchin) [reply core/io/pread_spec.rb:98]: I would suggest to use something other than `String` literal, e.g. `Object.new`. This way error message should be the same in both cases.

It's caused by using different C-functions to implicitly convert Ruby object to some type. One of them (`rb_num2ll` - to convert to `long long`) has a case for `String` object that leads to a specific error message. Other similar functions (to convert to `long` and `int`) don't have such case. Choosing what function to use depends on C numeric type sizes that is related to a compiler and C standard library implementation (and OS of course):

```c
#ifdef OFFT2NUM
# /* take that. */
#elif SIZEOF_OFF_T == SIZEOF_LONG_LONG
# define OFFT2NUM RB_LL2NUM
#elif SIZEOF_OFF_T == SIZEOF_LONG
# define OFFT2NUM RB_LONG2NUM
#else
# define OFFT2NUM RB_INT2NUM
#endif
```

**Comment by herwinw** [comment]
I believe all comments have been addressed now.

**Comment by andrykonchin** [comment]
Seems good! Could you please squash commits?

**Comment by andrykonchin** [comment]
Thank you!
